### ISSUE TITLE
Cannot create ObjectID for dates beyond 2038-01-19T03:14:07.000Z

### ISSUE THREAD

**Issue by williamkapke** [mongodb/js-bson#158]
The problem appears to be happening when the timestamp goes beyond `0x7fffffff`

```javascript
ObjectID(new Date(parseInt('7fffffff', 16)*1000).getTime()/1000).getTimestamp()
//Mon Jan 18 2038 19:14:07 GMT-0800 (PST)

ObjectID(new Date(parseInt('80000000', 16)*1000).getTime()/1000).getTimestamp()
//Fri Dec 13 1901 12:45:52 GMT-0800 (PST)
```

Same result when using `ObjectID.createFromTime()`.

**Comment by christkv** [comment]
This is a JS issue I think not the driver per say

**Comment by williamkapke** [comment]
You think this is a javascript issue?
```javascript
console.log(ObjectId("7fffffff0000000000000000").getTimestamp())
//Tue Jan 19 2038 03:14:07 GMT+0000 (UTC)
console.log(ObjectId("800000000000000000000000").getTimestamp())
//Fri Dec 13 1901 20:45:52 GMT+0000 (UTC)
```

This work perfectly fine in the [bson-objectid](https://www.npmjs.com/package/bson-objectid) module.

**Comment by christkv** [comment]
that's because the timestamp only uses 4 bytes for a seconds timestamp which means max 32 bit integer value is 0x7FFFFFFF.

**Comment by williamkapke** [comment]
Right- but why use a 32-bit *signed* integer? I mean, really this comes down to what's more important: Supporting `ObjectID`s created before 1970, or `ObjectID`s that go beyond 2038? I think companies would hope their database & code can continue to live on without this problem.

ANYHOW... I see this has been discussed long ago and the concern just isn't there :/
https://groups.google.com/forum/#!topic/mongodb-user/anfcl5cu8Xk

I guess this is still needing documentation.

**Comment by christkv** [comment]
Mongodb itself does not have unsigned integers.

You can define what those 12 bytes are and just make your own higher resolution timestamp but there needs to a uniqueness guarantee to use it as a _id value. 

https://docs.mongodb.org/manual/reference/object-id/

Another option is to use an uuid type key which removes a lot the constraints ObjectId has.

**Comment by williamkapke** [comment]
Some silly people (in very popular companies I won't mention) decided to leverage this timestamp "feature" and build code that relies on it. `ObjectID`s are being used all over the place now and for different reasons. AND... where that are used for `_id` there are queries that leverage the timestamps like: `{ _id: { $gt: ObjectId('5685c1806f00000000000000')} }` to find things like "Users that signed up after XX date)" (... which I know you know).

The bottom line is that people find this way of generating unique IDs pretty darn neeto and want to use it outside of mongo! I hope one day MongoDB Inc recognizes and embraces this, moves `ObjectID` out of `bson` and lets it grown on it's own.
